### PR TITLE
Refactor: useBlockTools hook

### LINKED ISSUE #57288: Central Command Area missing when Block with No Tools is Selected
### Description
When a block with no tools is selected, the top toolbar will show a collapse icon even though no block tools are there. I think we could put a `hasAnyBlockControls` check within the privately exported [`useCanBlockToolbarBeFocused`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/utils/use-can-block-toolbar-be-focused.js#L18).

### Step-by-step reproduction instructions
- Go to Site Editor
- Turn on Top Toolbar
- Edit Navigation Template Part
- Click navigation item
- Press Up Arrow until the block toolbar disappears and only a collapse icon shows in its place.

### Screenshots, screen recording, code snippet
Navigation Template Part with No Block Tools available and central document command center missing

### PR DESCRIPTION
**Pull request by jeryj**
Closes https://github.com/WordPress/gutenberg/issues/57288 alternative to https://github.com/WordPress/gutenberg/pull/57291 and https://github.com/WordPress/gutenberg/pull/58947.

There are times where a block is selected but no tools are available. If we don't have any tools, we don't want to show a toolbar. This combines the logic for showing the block toolbar that was sprinkled throughout the codebase:

- block-toolbar/index.js checked if it should render based on if a toolbar was enabled for the block and if there were block controls in the slots
- block-tools/index.js determined if the popover should render based on cascading rules if the inserter popover or block select mode popover were showing (and other rules)
- use-can-block-toolbar-be-focused made an approximation of if the block toolbar would be showing or not, but fell short in some areas such as when a block toolbar should have been rendered, but there were no block tools available.

## What?

Unify logic for which block tools will be showing so there is a source of truth for what will be showing (block toolbar popover, fixed block toolbar, block inserter, block breadcrumb).

## Why?

Fixes https://github.com/WordPress/gutenberg/issues/57288.

## How?

- Adds a `useHasBlockToolbar` hook that has the logic for if the block-toolbar will render or not.
- Adds a `useBlockTools` hook that privately exports which block tools will be showing (block toolbar popover, fixed block toolbar, block inserter, block breadcrumb).
- Checks if any block tools are available before adding the `is-collapsed` class to the central toolbar area
- Removes selected block tools wrapper from DOM if no block tools available. I'm not sure if this was intentional to leave it in the DOM before, or if it was because there wasn't a reliable way of checking for it.

## Testing Instructions

Toolbar should show/hide in the same way as before.

### Testing Instructions for Keyboard

- Go to Site Editor
- Turn on Top Toolbar
- Edit Navigation Template Part
- Click navigation item
- Press Up Arrow until the block toolbar disappears
- Central command area should be present
- Central command area should not have `is-collapsed` class
- Alt + F10 should move focus to the document tools

### CODE REVIEW

File: packages/block-editor/src/components/block-toolbar/index.js
Comment (draganescu) [inline packages/block-editor/src/components/block-toolbar/index.js]: Maybe
```suggestion
	const hasBlockToolbar = useHasBlockToolbar();
	if ( hasBlockToolbar ) {
```

It's the coding pattern used throughout the codebase.

File: packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js
Comment (draganescu) [inline packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js]:
```suggestion
 * Returns true if the block toolbar should be shown.
```

File: packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js
Comment (draganescu) [inline packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js:17]:
```suggestion
export function useShowBlockToolbar() {
```

File: packages/block-editor/src/components/block-tools/use-show-block-tools.js
Comment (draganescu) [inline packages/block-editor/src/components/block-tools/use-show-block-tools.js]: This comment and the name of the hook do not connect.

File: packages/edit-post/src/components/header/index.js
Comment (draganescu) [inline packages/edit-post/src/components/header/index.js]: could this come from a hook too?

File: packages/edit-site/src/components/header-edit-mode/index.js
Comment (draganescu) [inline packages/edit-site/src/components/header-edit-mode/index.js]: same comment as above, could this determination be abstracted in tiny hook?

**Comment by MaggieCabrera** [comment]
I think the label you want is Backport to WP beta/RC?

**Comment by jeryj** [comment]
Thanks @MaggieCabrera! I went ahead and removed the label because it is a larger change the bug it fixes is minor.

File: packages/block-editor/src/components/block-tools/use-show-block-tools.js
Comment (jeryj) [reply packages/block-editor/src/components/block-tools/use-show-block-tools.js]: Oh - thanks for the catch!

File: packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js
Comment (jeryj) [reply packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js:17]: This isn't a check to know if it will show or not, but rather if it can be shown. Like, does the current selection _have a block toolbar it can render_? `useShowBlockToolbar` makes me think it will do something to display the toolbar.

File: packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js
Comment (jeryj) [inline packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js:43]: This is the same check that `block-toolbar.js` used to determine if it would early return or continue to render.

File: packages/edit-site/src/components/header-edit-mode/index.js
Comment (jeryj) [reply packages/edit-site/src/components/header-edit-mode/index.js]: I refactored it further by adding `! isZoomOutMode` to the `showFixedToolbar` check and removing `hasFixedToolbar` here, as that setting was already checked in `showFixedToolbar`. Now the `hasTopToolbar` check is just `isLargeViewport && showFixedToolbar`. 

I don't think it's worth extracting that to its own hook as we can't easily share it between the `edit-site` and `edit-post` packages. Especially now that it's just an `isLargeViewport` check. I didn't want to put the `showTopToolbar` vs `showFixedToolbar` within the `block-editor` package, as it isn't supposed to have knowledge about how the block toolbar is being used outside of it.

File: packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js
Comment (Mamaduka) [inline packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js:28]: Suggestion: We can probably use `getBlockSelectionStart` here if we want only the first block clientId.

File: packages/block-editor/src/components/block-tools/use-show-block-tools.js
Comment (Mamaduka) [inline packages/block-editor/src/components/block-tools/use-show-block-tools.js:68]: This hook subscribes to four values. Three of these are only used by `BlockTools` and the remaining `showFixedToolbar` by only editor packages.

I'm unsure if moving the selection of values used only by `BlockTools` to this hook is worth it. cc @ellatrix

File: packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js
Comment (jeryj) [reply packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js:28]: I do think that would be a good change because `getBlockSelectionStart()` returns a state value and `selectedBlockClientIds()` does quite a bit more work. However, if I:

- Add three paragraphs
- click and drag from the third paragraph up to the first
- `getBlockSelectionStart()` will return the third paragraph ID (where selection started) and  `getSelectedBlockClientIds()[0]` will return the first paragraph ID.

So, as it would be a functional change, I think we should do it in a follow-up.

File: packages/block-editor/src/components/block-tools/use-show-block-tools.js
Comment (jeryj) [reply packages/block-editor/src/components/block-tools/use-show-block-tools.js:68]: I had concerns about that too. However, the information we need for `showFixedToolbar` is all right here, and this hook is already running for `<BlockTools />`. It doesn't need to do any additional work to set this value. I think it will end up being  more performant and less code, so it's worth the tradeoff.

File: packages/block-editor/src/components/block-tools/use-show-block-tools.js
Comment (Mamaduka) [inline packages/block-editor/src/components/block-tools/use-show-block-tools.js]: The `isUnmodifiedDefaultBlock` expects the block object. Even though it currently only uses `attribute` and `name` properties, I think this is more future proof.

Other changes are just minor cleanups.

**Review by Mamaduka — APPROVED** [review]
@jeryj, the refactoring looks good, and the changes test well for me across the editors.

Approving the PR, but it would be nice to get a second opinion before merging, just in case I missed something 😅

P.S. Can you rebase the changes on top of the latest trunk?

File: packages/block-editor/src/components/block-tools/use-show-block-tools.js
Comment (Mamaduka) [reply packages/block-editor/src/components/block-tools/use-show-block-tools.js:68]: This is probably fine. We can always refactor if [CodeVitals](https://www.codevitals.run/project/gutenberg) catches any perf regression.

File: packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js
Comment (jeryj) [reply packages/block-editor/src/components/block-toolbar/use-has-block-toolbar.js:28]: Follow-up at https://github.com/WordPress/gutenberg/pull/59450

File: packages/block-editor/src/components/block-tools/use-show-block-tools.js
Comment (jeryj) [reply packages/block-editor/src/components/block-tools/use-show-block-tools.js]: Follow up at https://github.com/WordPress/gutenberg/pull/59450

**Comment by youknowriad** [comment]
Hi there, it looks like this PR introduced a meaningful performance regression for the "block select" metric. Can we look into it?

**Comment by jeryj** [comment]
@youknowriad Sure! Thanks for flagging it.

**Comment by jeryj** [comment]
@youknowriad I'm looking at https://www.codevitals.run/project/gutenberg and I see the increased Block Select time, but I'm having a hard time tracking down what this metric is really measuring. I've been looking through the e2e performance tests and Readme.  In the performance test for this branch, [the one that stands out as an increase is `firstPaint`](https://github.com/WordPress/gutenberg/actions/runs/8084710994/job/22090588993), which seems different than Block Select. 

Could you point me in the right direction for debugging this more?

**Comment by youknowriad** [comment]
Yes, it's this one https://github.com/WordPress/gutenberg/blob/f3676dcb33ac2b4a57ef2b64ca6fcf96b91d9683/test/performance/specs/post-editor.spec.js#L249-L297 it's basically the sum of the "focus" + "focusin" event handlers when you click on an unselected paragraph block.

File: packages/block-editor/src/components/block-tools/use-show-block-tools.js
Comment (ellatrix) [reply packages/block-editor/src/components/block-tools/use-show-block-tools.js:68]: It's fine, it's not worth trying to optimise performance here because there's only one block toolbar visible at any point in time.

Comment (ellatrix) [reply packages/block-editor/src/components/block-tools/use-show-block-tools.js:68]: Or maybe not, looking at the comments below. Not sure if this is the cause of the regression though.

**Comment by jeryj** [comment]
@youknowriad @ellatrix The performance regression is because in order to know if the block toolbar will render or not, we need to rely on the [`useHasAnyBlockControls`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-controls/use-has-block-controls.js#L12) hook, which loops through each slot in the block toolbar to check if there's a fill, and can't return early since it's a hook.  

This was seemingly good enough when it was only within the `<BlockToolbar />` component, but this PR intended to have one hook that could be used anywhere to tell if the block toolbar would be rendered or not. That means relying on `useHasAnyBlockControls`. Unfortunately, when this is used as a hook dependency, it starts to get expensive.

So, we could explore:
- Find a different way to check if the block toolbar will render (not rely on checking all slots for a fill?). It seems like other approaches were explored before https://github.com/WordPress/gutenberg/pull/53410 landed.
- Adding an `isBlockToolbarAvailable` state to the block editor store so we don't need to rely on this hook
- … I'll keep thinking on it. Very open to ideas!

**Comment by ellatrix** [comment]
Why do we need to check fills? Don't we always need to render the switcher and options buttons?

**Comment by jeryj** [comment]
@ellatrix There are a few very specific instances:

- [The navigation block in isolated editor](https://github.com/WordPress/gutenberg/pull/53110): You can get to a point where you have a block selected but no block tools available
- The featured image block in content only mode offers a Replace button to replace the image. 

https://github.com/WordPress/gutenberg/pull/53410 explains it well. 

I'm not sure if those things are doing it wrong and should be fixed in those locations, or if they have valid use cases that we need to support here.

**Comment by ellatrix** [comment]
Not sure I get this. To me the should always be the block icon visible?

For this PR specifically, can't you check first if there is anything else to render in the toolbar before making this expensive check? If there's something to render, there's no need to do this check.

**Comment by jeryj** [comment]
The featured image replace one specifically is when the editor is in contentOnly mode, and the toolbar mostly only renders things in default editor mode. I'm not saying this is a good idea, but that's how it's set up. 

I'd love to get rid of the code that checks all the block tools slots for fills. It seems like a heavy solution to render one button in one mode for one block that affects the performance of everything else.

**Comment by ellatrix** [comment]
Could you still do other checks first to fix the performance problem?

**Comment by youknowriad** [comment]
Any follow-ups here, let's not let it slip

**Comment by jeryj** [comment]
@youknowriad - I'm feeling stuck on how to address `useHasAnyBlockControls`. I spent a bit of time poking at different ways to improve the performance, and `useHasAnyBlockControls` was the main culprit. It's only necessary for a couple of [very specific instances](https://github.com/WordPress/gutenberg/pull/58979#issuecomment-1971785874) that I think may be misusing the intention of the block toolbar. If we could change those implementations, we can remove `useHasAnyBlockTools` entirely.

Previously, `useHasAnyBlockControls` wouldn't run as often because it was the last check before rendering the toolbar. This PR lifted it to a hook that could be used anywhere to know if the block toolbar _would_ render or not. I still think this is a useful thing for us and third party tools. 

https://github.com/WordPress/gutenberg/pull/58979#issuecomment-1971681017 and https://github.com/WordPress/gutenberg/pull/53410 explain `useHasAnyBlockControls` more. Very open to direction, as I'm feeling a bit stuck on it.

**Comment by youknowriad** [comment]
Is it possible to keep the call to `useHasAnyBlockControls` where it was before while we solve these very specific instances (remove the need for the check for these cases)?

**Comment by jeryj** [comment]
> Is it possible to keep the call to `useHasAnyBlockControls` where it was before 

Unfortunately not :/ The purpose of this PR was to make it possible to know if the block toolbar would render, and leaving `useHasAnyBlockControls` where it was before would make this refactor no longer reliable.

**Comment by jeryj** [comment]
If we address [the content-locked image block toolbar offering a "replace" button](https://github.com/WordPress/gutenberg/pull/53410), then we could go back to [checking for content-locked](https://github.com/WordPress/gutenberg/pull/53110) to not render the toolbar again. This would allow us to remove the `useHasAnyBlockControls` check, which was added only for the image and featured image blocks:.

**Comment by Mamaduka** [comment]
@jeryj, both links are pointing to the same pull request.

**Comment by youknowriad** [comment]
To be honest, the bug fixed by this PR feels rather small while performance regression are experienced by everyone. I'm not saying we should not fix it but maybe it help prioritize things.

**Comment by jeryj** [comment]
> @jeryj, both links are pointing to the same pull request.

Updated. Thanks for catching, @Mamaduka!

> the bug fixed by this PR feels rather small while performance regression are experienced by everyone.

Fair. I'm putting in a PR that _only_ checks for the "other" slot, as that's the only one used. It reduces the selecting block focus by 3-4ms. It's not a perfect code solution, but it does reduce the impact.

**Comment by jeryj** [comment]
Performance regression fixed by https://github.com/WordPress/gutenberg/pull/60717 and https://github.com/WordPress/gutenberg/pull/60879. Now we're below where the block select metric started: 
Block select metric showing the initial rise to from 30ms to 45ms average and now back down to upper 20ms

**Comment by youknowriad** [comment]
Great thanks for the follow-ups :)

**Pull request merged**
